batchlog_manager: set timeout on replay of a failed batch

mlitvk opened the pull request:
When replaying a failed batch and sending the mutation to all replicas,
set a reasonable timeout on the request.

Without a timeout, if a replica does not respond, the batchlog manager's
replay loop can become stuck waiting indefinitely. This prevents any
further failed batches from being applied until the node is restarted,
and can also cause node shutdown to hang while waiting for the batchlog
loop to complete.

backport to relevant versions since the issue was observed cause node shutdown to hang

mlitvk commented on service/storage_proxy.cc:
```diff
@@ -4057,6 +4057,7 @@ storage_proxy::mutate_atomically_result(std::vector<mutation> mutations, db::con
         };
         future<> async_remove_from_batchlog() {
             // delete batch
+            utils::get_local_injector().inject("storage_proxy_fail_remove_from_batchlog", [] { throw std::runtime_error("Error injection: failing remove from batchlog"); });
```
I checked and it should be optimized away in release, thanks

mlitvk commented on db/batchlog_manager.cc:
```diff
@@ -116,7 +116,8 @@ future<> db::batchlog_manager::batchlog_replay_loop() {
         } catch (...) {
             blogger.error("Exception in batch replay: {}", std::current_exception());
         }
```
the normal interval of the replay loop is 60 seconds. the injection changes it to run every 1 seconds to make the test run faster. I think it's clear enough

mykaul commented:
Is there an issue this fixes/refs?

mlitvk commented:
> Is there an issue this fixes/refs?

it's related to https://github.com/scylladb/scylla-enterprise/issues/5499

I wasn't sure since it's an enterprise issue. should I open a new issue in this repo?

mykaul commented:
> > Is there an issue this fixes/refs?
> 
> it's related to [scylladb/scylla-enterprise#5499](https://github.com/scylladb/scylla-enterprise/issues/5499)
> 
> I wasn't sure since it's an enterprise issue. should I open a new issue in this repo?

Yes please, with distilled details.

piodul commented on db/batchlog_manager.cc:
```diff
@@ -242,7 +242,8 @@ future<> db::batchlog_manager::replay_all_failed_batches(post_replay_cleanup cle
                 // have hints (yet), send with CL=ALL, and hope we can re-do this soon.
                 // See below, we use retry on write failure.
-                return _qp.proxy().mutate(mutations, db::consistency_level::ALL, db::no_timeout, nullptr, empty_service_permit(), db::allow_per_partition_rate_limit::no);
+                auto timeout = db::timeout_clock::now() + _write_request_timeout;
+                return _qp.proxy().mutate(mutations, db::consistency_level::ALL, timeout, nullptr, empty_service_permit(), db::allow_per_partition_rate_limit::no);
```
The problem with using a timeout here is that it's not clear what the timeout should be. Imagine that you configured the node with a small write timeout and writing the batch initially fails due to write timeout - because it was just big and applying it took a lot of time. IIUC, the retry logic here will try to use the same value for the timeout here, so in this particular case batchlog replay will get stuck.

I'm not really sure what should be a good value for the timeout here.

We had the same issues with respect to hint replay in the past and IIRC the solution to the problem was to abort in-flight hint writes when shutting down. How difficult would it be to implement this for batch writes?

mykaul replied:
@denesb may have an idea.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17919/) | :white_check_mark: | i-0610c2d4200aa985c |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17919/) | :white_check_mark: | i-0924be0eed5e1e215 |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17919/) | :white_check_mark: | i-0924be0eed5e1e215 |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17919/)<br> :small_blue_diamond: cluster/test_batchlog_manager | :white_check_mark: | i-0924be0eed5e1e215 |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17919/) | :white_check_mark: | i-09ded17effc1b94c5 |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17919/) | :white_check_mark: | i-09ded17effc1b94c5 |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17919/)<br> :small_blue_diamond: cluster/test_batchlog_manager | :white_check_mark: | i-09ded17effc1b94c5 |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17919/) | :white_check_mark: | i-034d47517f8efe5d6 |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17919/) | :white_check_mark: | i-034d47517f8efe5d6 |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17919/)<br> :small_blue_diamond: cluster/test_batchlog_manager | :white_check_mark: | i-034d47517f8efe5d6 |
| dtest | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16988/) | :white_check_mark: |  |
|  | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8902/) | :white_check_mark: |  |
|  | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5717/) | :white_check_mark: |  |

#### Build Details:
- Duration: 2 hr 23 min

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17922/) | :white_check_mark: | i-08c7b3a66ff370a75 |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17922/) | :white_check_mark: | i-0fe7dd04122c7c1fe |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17922/) | :white_check_mark: | i-0fe7dd04122c7c1fe |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17922/)<br> :small_blue_diamond: cluster/test_batchlog_manager | :white_check_mark: | i-0fe7dd04122c7c1fe |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17922/) | :white_check_mark: | i-0819b5b319ffa8b64 |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17922/) | :white_check_mark: | i-0819b5b319ffa8b64 |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17922/)<br> :small_blue_diamond: cluster/test_batchlog_manager | :white_check_mark: | i-0819b5b319ffa8b64 |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17922/) | :white_check_mark: | i-0561c04822de5f8cb |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17922/) | :white_check_mark: | i-0561c04822de5f8cb |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17922/)<br> :small_blue_diamond: cluster/test_batchlog_manager | :white_check_mark: | i-0561c04822de5f8cb |
| dtest | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8906/) | :white_check_mark: |  |
|  | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16992/) | :white_check_mark: |  |
|  | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5718/) | :white_check_mark: |  |

#### Build Details:
- Duration: 2 hr 2 min

xemul commented on db/batchlog_manager.hh:
```diff
@@ -45,6 +45,7 @@ public:
 private:
     static constexpr uint32_t replay_interval = 60 * 1000; // milliseconds
     static constexpr uint32_t page_size = 128; // same as HHOM, for now, w/out using any heuristics. TODO: set based on avg batch size.
+    static constexpr uint32_t write_timeout_in_seconds = 300;
```
nit: worth having it as std::chrono::seconds right here, you convert it into seconds by the caller anyway

mlitvk replied:
I changed the type to chrono::seconds, and changed also the type of `replay_interval` while at it

mlitvk commented on service/storage_proxy.cc:
```diff
@@ -4252,6 +4268,15 @@ future<> storage_proxy::send_hint_to_all_replicas(frozen_mutation_and_schema fm_
+            return batchlog_replay_mutation(std::move(m));
+        }) | std::ranges::to<std::vector<batchlog_replay_mutation>>();
+
+    return mutate_internal(std::move(ms), db::consistency_level::ALL, nullptr, empty_service_permit(), timeout)
```
as far as I understand, in the normal write path in `mutate_atomically_result` the first thing we do is cdc augmentation, then write it to the batchlog and send it to all replicas.
so I think we can't have mutations in the batchlog that are not cdc augmented.

xemul replied:
Is it possible to cook a test that validates that? The fault injection can be of great help. Or maybe we already have such.

mlitvk replied:
I think it's not written to the batchlog.
it has type `COUNTER` so it's not atomic:
```
    if (_type != type::LOGGED) {                                                                                                                                                                                                              
        _stats.batches_pure_unlogged += 1;                                                                                                                                                                                                    
        mutate_atomic = false;           
    } else {                                                                                                                                                                                                                                                                                                                                                                                                                                       
...
    }                                                                                                                                                                                                                                         
    return qp.proxy().mutate_with_triggers(std::move(mutations), cl, timeout, mutate_atomic, std::move(tr_state), std::move(permit), db::allow_per_partition_rate_limit::yes);                                                                
```
and `mutate_with_triggers` goes to the normal mutate path, and not the atomic path where we write to the batchlog

mlitvk commented:
> Now that ci has passed are there any more review comments to resolve? If not, @mlitvk please label the PR as merge candidate and ping `@scylladb/scylla-maint`

currently I addressed all comments and I'm waiting for approval from the reviewers or more comments

bhalevy commented:
@xemul / @piodul please approve and let's push this forward.  We seem to have dropped the ball on it...

piodul commented on test/cluster/test_batchlog_manager.py:
```diff
@@ -0,0 +1,131 @@
+            # injected error is expected
+            logger.error(f"Error executing batch: {e}")
+
+        await asyncio.gather(*[manager.api.disable_injection(s.ip_addr, "storage_proxy_fail_remove_from_batchlog") for s in servers])
+
+        await manager.server_stop(servers[1].server_id)
```
Same problem with shutdown vs. injection disabling racing here.

piodul commented on test/cluster/test_batchlog_manager.py:
```diff
@@ -129,3 +129,60 @@ async def batchlog_empty() -> bool:
+            result2 = await cql.run_async(f"SELECT * FROM {cdc_table_name} WHERE key = 40 ALLOW FILTERING")
+            if len(result1) > 0 and len(result2) > 0:
+                return True
+        await wait_for(cdc_data_present, time.time() + 60)
```
How about, after this line, adding a check that the number of rows is as expected? Specifically, if the wrong `mutate`function is used by batchlog replay then the CDC log mutations might be generated for a second time and the result will contain duplicated rows (AFAIK the rows could get a different cdc$time, which is a timeuuid and contains a random part).

mlitvk commented on test/cluster/test_batchlog_manager.py:
```diff
@@ -0,0 +1,131 @@
+            # injected error is expected
+            logger.error(f"Error executing batch: {e}")
+
+        await asyncio.gather(*[manager.api.disable_injection(s.ip_addr, "storage_proxy_fail_remove_from_batchlog") for s in servers])
+
+        await manager.server_stop(servers[1].server_id)
```
fixed, thanks. I added an injection `skip_batch_replay` that skips batch replay, and I use it to enable batch replay only after the node is stopped
